File: python/foundry-local-whisper.py
"""
Whisper Voice Transcription with Foundry Local
Transcribes WAV audio files using the OpenAI Whisper model running locally.

Uses the Foundry Local SDK to download and manage the whisper model, then
runs inference directly with ONNX Runtime and the transformers feature
extractor. Requires: foundry-local-sdk, onnxruntime, transformers, librosa

Usage: python foundry-local-whisper.py [path-to-wav-file]
"""

import sys
import os
import glob
import time

import numpy as np
import onnxruntime as ort
import librosa
from transformers import WhisperFeatureExtractor, WhisperTokenizer
from foundry_local import FoundryLocalManager

model_alias = "whisper-medium"

# Default: transcribe all Zava sample WAV files
samples_dir = os.path.join(os.path.dirname(__file__), "..", "samples", "audio")

if len(sys.argv) > 1:
    audio_files = [sys.argv[1]]
    for f in audio_files:
        if not os.path.exists(f):
            print(f"Audio file not found: {f}")
            print("Usage: python foundry-local-whisper.py [path-to-wav-file]")
            sys.exit(1)
else:
    audio_files = sorted(glob.glob(os.path.join(samples_dir, "zava-*.wav")))
    if not audio_files:
        print(f"No WAV files found in {samples_dir}")
        print("Run 'python samples/audio/generate_samples.py' first to generate them.")
        sys.exit(1)

# ---------------------------------------------------------------------------
# Step 1: Use Foundry Local SDK to ensure the whisper model is downloaded
# ---------------------------------------------------------------------------
print(f"Initialising Foundry Local with model: {model_alias}...")
manager = FoundryLocalManager(model_alias)
model_info = manager.get_model_info(model_alias)
cache_location = manager.get_cache_location()

# Build the path to the cached ONNX model files
model_dir = os.path.join(
    cache_location, "Microsoft",
    model_info.id.replace(":", "-"),
    "cpu-fp32"
)

if not os.path.isdir(model_dir):
    print(f"Model directory not found: {model_dir}")
    print("Ensure the whisper model has been downloaded by the SDK.")
    sys.exit(1)

print(f"Model ready: {model_info.id}")
print(f"Model path: {model_dir}")

# ---------------------------------------------------------------------------
# Step 2: Load the encoder and decoder ONNX sessions
# ---------------------------------------------------------------------------
print("\nLoading ONNX encoder and decoder...")
encoder_session = ort.InferenceSession(
    os.path.join(model_dir, "whisper-medium_encoder_fp32.onnx"),
    providers=["CPUExecutionProvider"],
)
decoder_session = ort.InferenceSession(
    os.path.join(model_dir, "whisper-medium_decoder_fp32.onnx"),
    providers=["CPUExecutionProvider"],
)

# Load the feature extractor and tokeniser from the model directory
feature_extractor = WhisperFeatureExtractor.from_pretrained(model_dir)
tokenizer = WhisperTokenizer.from_pretrained(model_dir)

# Whisper decoder dimensions (medium model)
NUM_LAYERS = 24
NUM_HEADS = 16
HEAD_SIZE = 64

# Build the initial decoder token sequence:
# <|startoftranscript|> <|en|> <|transcribe|> <|notimestamps|>
sot = tokenizer.convert_tokens_to_ids("<|startoftranscript|>")
eot = tokenizer.convert_tokens_to_ids("<|endoftext|>")
notimestamps = tokenizer.convert_tokens_to_ids("<|notimestamps|>")
forced_ids = tokenizer.get_decoder_prompt_ids(language="en", task="transcribe")
INITIAL_TOKENS = [sot] + [tid for _, tid in forced_ids] + [notimestamps]

print("Models loaded and ready.\n")


def transcribe(audio_path: str) -> str:
    """Transcribe a single WAV file and return the text."""
    # Load audio at 16 kHz mono
    audio, _ = librosa.load(audio_path, sr=16000)

    # Extract log-mel spectrogram features
    features = feature_extractor(audio, sampling_rate=16000, return_tensors="np")
    audio_features = features["input_features"].astype(np.float32)

    # Run the encoder
    encoder_outputs = encoder_session.run(None, {"audio_features": audio_features})
    cross_kv_list = encoder_outputs[1:]

    # Prepare cross-attention KV cache from encoder
    cross_kv = {}
    for i in range(NUM_LAYERS):
        cross_kv[f"past_key_cross_{i}"] = cross_kv_list[i * 2]
        cross_kv[f"past_value_cross_{i}"] = cross_kv_list[i * 2 + 1]

    # Initialise empty self-attention KV cache
    self_kv = {}
    for i in range(NUM_LAYERS):
        self_kv[f"past_key_self_{i}"] = np.zeros((1, NUM_HEADS, 0, HEAD_SIZE), dtype=np.float32)
        self_kv[f"past_value_self_{i}"] = np.zeros((1, NUM_HEADS, 0, HEAD_SIZE), dtype=np.float32)

    # Autoregressive decoding
    input_ids = np.array([INITIAL_TOKENS], dtype=np.int32)
    generated = []

    for _ in range(448):
        feeds = {"input_ids": input_ids}
        feeds.update(cross_kv)
        feeds.update(self_kv)

        outputs = decoder_session.run(None, feeds)
        logits = outputs[0]
        next_token = int(np.argmax(logits[0, -1, :]))

        if next_token == eot:
            break

        generated.append(next_token)

        # Update self-attention KV cache
        for i in range(NUM_LAYERS):
            self_kv[f"past_key_self_{i}"] = outputs[1 + i * 2]
            self_kv[f"past_value_self_{i}"] = outputs[2 + i * 2]

        input_ids = np.array([[next_token]], dtype=np.int32)

    return tokenizer.decode(generated, skip_special_tokens=True).strip()


# ---------------------------------------------------------------------------
# Step 3: Transcribe each audio file
# ---------------------------------------------------------------------------
for audio_path in audio_files:
    filename = os.path.basename(audio_path)
    print(f"{'=' * 60}")
    print(f"File: {filename}")
    print("=" * 60)

    t0 = time.time()
    text = transcribe(audio_path)
    elapsed = time.time() - t0

    print(text)
    print(f"({elapsed:.1f}s)\n")

print(f"Done — transcribed {len(audio_files)} file(s).")

File: javascript/foundry-local-whisper.mjs
/**
 * Whisper Voice Transcription with Foundry Local
 * Transcribes WAV audio files using the OpenAI Whisper model running locally.
 *
 * Uses the Foundry Local SDK's built-in AudioClient for transcription,
 * which handles all ONNX inference and audio preprocessing internally.
 *
 * Requires: foundry-local-sdk
 * Usage: node foundry-local-whisper.mjs [path-to-wav-file]
 */

import { FoundryLocalManager } from "foundry-local-sdk";
import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const modelAlias = "whisper-medium";

// ---- Audio file discovery ------------------------------------------------

const samplesDir = path.join(__dirname, "..", "samples", "audio");
let audioFiles;

if (process.argv[2]) {
  const filePath = process.argv[2];
  if (!fs.existsSync(filePath)) {
    console.error(`Audio file not found: ${filePath}`);
    process.exit(1);
  }
  audioFiles = [filePath];
} else {
  if (!fs.existsSync(samplesDir)) {
    console.error(`Samples directory not found: ${samplesDir}`);
    process.exit(1);
  }
  audioFiles = fs
    .readdirSync(samplesDir)
    .filter((f) => f.startsWith("zava-") && f.endsWith(".wav"))
    .sort()
    .map((f) => path.join(samplesDir, f));
  if (audioFiles.length === 0) {
    console.error(`No WAV files found in ${samplesDir}`);
    process.exit(1);
  }
}

// ---- Step 1: Start Foundry Local and prepare the model -------------------

console.log(`Initialising Foundry Local with model: ${modelAlias}...`);
FoundryLocalManager.create({ appName: "FoundryLocalWorkshop" });
const manager = FoundryLocalManager.instance;
await manager.startWebService();

const catalog = manager.catalog;
const model = await catalog.getModel(modelAlias);

if (!model.isCached) {
  console.log(`Downloading model: ${modelAlias} (this may take several minutes)...`);
  await model.download();
  console.log(`Download complete: ${modelAlias}`);
}

await model.load();
console.log(`Model ready: ${model.id}`);

// ---- Step 2: Create an AudioClient for transcription ---------------------

const audioClient = model.createAudioClient();
audioClient.settings.language = "en";
console.log("Audio client ready.\n");

// ---- Helper: split a WAV file into ≤30-second chunk files ----------------
// Whisper's encoder processes exactly 30 seconds per pass. For longer files
// we split the PCM data, write temporary WAV files, and transcribe each.
// See https://github.com/microsoft/Foundry-Local/issues/517

const CHUNK_SECONDS = 30;

function splitWavIntoChunks(wavPath) {
  const buf = fs.readFileSync(wavPath);
  // Parse WAV header (standard 44-byte RIFF/PCM header)
  const numChannels = buf.readUInt16LE(22);
  const sampleRate = buf.readUInt32LE(24);
  const bitsPerSample = buf.readUInt16LE(34);
  const bytesPerSample = (bitsPerSample / 8) * numChannels;
  const headerSize = 44;
  const pcmData = buf.subarray(headerSize);
  const totalSamples = pcmData.length / bytesPerSample;
  const chunkSamples = CHUNK_SECONDS * sampleRate;

  if (totalSamples <= chunkSamples) return null; // no splitting needed

  const numChunks = Math.ceil(totalSamples / chunkSamples);
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "whisper-chunks-"));
  const chunkPaths = [];

  for (let i = 0; i < numChunks; i++) {
    const startByte = i * chunkSamples * bytesPerSample;
    const endByte = Math.min(startByte + chunkSamples * bytesPerSample, pcmData.length);
    const chunkPcm = pcmData.subarray(startByte, endByte);
    const chunkLen = chunkPcm.length;

    // Build a new WAV header for this chunk
    const header = Buffer.alloc(headerSize);
    header.write("RIFF", 0);
    header.writeUInt32LE(36 + chunkLen, 4);
    header.write("WAVE", 8);
    header.write("fmt ", 12);
    header.writeUInt32LE(16, 16); // fmt chunk size
    header.writeUInt16LE(1, 20);  // PCM format
    header.writeUInt16LE(numChannels, 22);
    header.writeUInt32LE(sampleRate, 24);
    header.writeUInt32LE(sampleRate * bytesPerSample, 28);
    header.writeUInt16LE(bytesPerSample, 32);
    header.writeUInt16LE(bitsPerSample, 34);
    header.write("data", 36);
    header.writeUInt32LE(chunkLen, 40);

    const chunkPath = path.join(tmpDir, `chunk-${i}.wav`);
    fs.writeFileSync(chunkPath, Buffer.concat([header, chunkPcm]));
    chunkPaths.push(chunkPath);
  }

  return { chunkPaths, tmpDir, numChunks, duration: totalSamples / sampleRate };
}

// ---- Step 3: Transcribe each audio file ----------------------------------

for (const audioPath of audioFiles) {
  const filename = path.basename(audioPath);
  console.log("=".repeat(60));
  console.log(`File: ${filename}`);
  console.log("=".repeat(60));

  const t0 = Date.now();
  const chunks = splitWavIntoChunks(audioPath);

  let text;
  if (!chunks) {
    // Short audio (≤30 s) — single pass
    const result = await audioClient.transcribe(audioPath);
    text = result.text;
  } else {
    // Long audio — transcribe each chunk and concatenate
    console.log(`  Audio is ${chunks.duration.toFixed(1)}s — splitting into ${chunks.numChunks} chunks of ${CHUNK_SECONDS}s`);
    const parts = [];
    for (let i = 0; i < chunks.chunkPaths.length; i++) {
      const chunkDur = Math.min(CHUNK_SECONDS, chunks.duration - i * CHUNK_SECONDS);
      process.stdout.write(`  Chunk ${i + 1}/${chunks.numChunks} (${chunkDur.toFixed(1)}s)... `);
      const result = await audioClient.transcribe(chunks.chunkPaths[i]);
      console.log("done");
      if (result.text) parts.push(result.text);
    }
    text = parts.join(" ");
    // Clean up temp files
    for (const p of chunks.chunkPaths) fs.unlinkSync(p);
    fs.rmdirSync(chunks.tmpDir);
  }

  const elapsed = ((Date.now() - t0) / 1000).toFixed(1);
  console.log(text);
  console.log(`(${elapsed}s)\n`);
}

console.log(`Done — transcribed ${audioFiles.length} file(s).`);

// Cleanup: unload the model to release resources
await model.unload();
await manager.stopWebService();
